### DISCUSSION TITLE
Build error: Nextjs 13 with Typescript and app/ experimental.

### DISCUSSION THREAD

**Discussion by jercomio** [vercel/next.js#46275] [Help] [answered]
My App run in dev mod but not when I build it. Somebody have an idea why ?
I use Tailwindcss and Postcss.

> next build
warn  - You have enabled experimental features (appDir, mdxRs) in next.config.js.
warn  - Experimental features are not covered by semver, and may cause unexpected or broken application behavior. Use at your own risk.
info  - Thank you for testing `appDir` please leave your feedback at https://nextjs.link/app-feedback

info  - Creating an optimized production build  
info  - Compiled successfully

info  - Linting and checking validity of types  
info  - Collecting page data  
[    ] info  - Generating static pages (0/3)ReferenceError: document is not defined
at JrcLogo (.next/server/app/page.js:913:19)
    at ae (node_modules/next/dist/compiled/react-dom/cjs/react-dom-server.browser.production.min.js:99:272)
    at ae (node_modules/next/dist/compiled/react-dom/cjs/react-dom-server.browser.production.min.js:104:142)
    at Z (node_modules/next/dist/compiled/react-dom/cjs/react-dom-server.browser.production.min.js:105:91)
    at be (node_modules/next/dist/compiled/react-dom/cjs/react-dom-server.browser.production.min.js:107:98)
    at de (node_modules/next/dist/compiled/react-dom/cjs/react-dom-server.browser.production.min.js:106:424)
    at Z (node_modules/next/dist/compiled/react-dom/cjs/react-dom-server.browser.production.min.js:105:222)
    at be (node_modules/next/dist/compiled/react-dom/cjs/react-dom-server.browser.production.min.js:107:98)
    at ae (node_modules/next/dist/compiled/react-dom/cjs/react-dom-server.browser.production.min.js:100:213)
    at Z (node_modules/next/dist/compiled/react-dom/cjs/react-dom-server.browser.production.min.js:105:91)

Error occurred prerendering page "/". Read more: https://nextjs.org/docs/messages/prerender-error
ReferenceError: document is not defined

Build error occurred
Error: Export encountered errors on following paths:
        /page: /
    at /node_modules/next/dist/export/index.js:415:19
    at processTicksAndRejections (node:internal/process/task_queues:96:5)

**Comment by icyJoseph** [comment, marked as answer]
A repository would be helpful.

With the given information I see:

```console
ReferenceError: document is not defined
```

This tells me that somewhere in your `/` page, or any of its children, you are accessing `document`, for example with `document.getElementById`, or any other document API, without using proper guards/checks.

During SSR, in Pages Dir, or App Dir, there's no window, nor document, nor DOM, so you cannot try to read from these APIs. Why are these API's missing? Because they belong to the client/browser runtime. 

> I guess you have also been told that Client Components only run client side, which is not correct. They are part of the SSR process, and just like with Pages, you gotta be careful accessing window/DOM

A safe place to access window or DOM APIs is within useEffect. Again without a repository I cannot help you more than this.

**Reply by jercomio** [reply]
Thanks !
If you can help me, look at the repository: https://github.com/jercomio/examples

**Reply by icyJoseph** [reply]
These lines for example, https://github.com/jercomio/examples/blob/main/src/components/DevCard/Elements/FrameworksBar.tsx#L14-L16:

```js
const FrameworksBar = () => {
  const [mouseOver, setMouseOver] = useState(false);

  const tabBar = document.querySelector(".tab-bar");
  const tab = document.getElementById("tab");
  const linksTarget = document.querySelectorAll("a");
```

That ain't gonna fly. `document` is not available during SSR.

There's this conceptual error, or slightly misinformation going around, that, client components (with "use client") only run client side, that's not correct, within Next.js at least. I don't know if other frameworks want to do that, but it feels wrong to not render them server side. Why? Because the HTML associated with these components wouldn't be part of the HTML seen by bots or people with JavaScript disabled.

It is better to use `ref`'s for what you are trying to gain access to the DOM nodes created by React:

- https://beta.reactjs.org/reference/react/useRef
- https://beta.reactjs.org/reference/react/useRef#manipulating-the-dom-with-a-ref

**Reply by jercomio** [reply]
Ok, I'll try... thank you so much !

**Reply by icyJoseph** [reply]
Let me know if you get stuck or something is not clear. I feel like you are in the right path!

**Reply by jercomio** [reply]
Well, I updated repository. I fixed the errors, no typescript errors but the components JrcLogo and FrameworksBar doesn't work. Certainly, I must have wrote it wrong...

**Reply by icyJoseph** [reply]
I see you are still using query selector, on tab-bar for example, but you don't need that, you already have the element on the ref.

**Reply by jercomio** [reply]
Ah yes, I believe that I have understood...

**Reply by icyJoseph** [reply]
Did it work?

**Reply by jercomio** [reply]
I updated the repo. It work but for FrameworksBar component there is a little problem when the cursor is over icon.

**Reply by icyJoseph** [reply]
Ah well, that sounds like a React issue. I might take a look later. As far as the document error, we are clear right?

It's a common pitfall of switching from client side rendering to SSR. 

Don't forget to mark as answered :) - if not today, tomorrow I can take a look at your hover issue.

**Reply by jercomio** [reply]
Done, the repo is updated and now the components works ! And when I run build App, I've no errors.
So, thank you very much for your help. Great !
